### PR TITLE
Disabling users in WebUI

### PR DESCRIPTION
**Pull request by Faldon**
# Ability to disable users in Web UI
## Base for implementing #283 
- Implemented an action menu for each user like in the files section
- Added the menu items delete and disable/enable (based on the current state of the user)
- Added a Disabled Group in the sidebar and a disbaled users count
## TODO:
- ~~When enabling a user, site needs a reload because the groups won't be filled with users anymore.
  The other way around seems to work fine.~~ (Edit: I just fired ab my local test server after the weekend and this strange behaviour was gone. Maybe a cache problem?)

### CODE REVIEW

**Comment by MorrisJobke** [comment]
* [ ] PNGs are not needed anymore on master
* [ ] optimize the SVGs (use either `svgo` or `scour` to do so)

File: settings/Controller/UsersController.php [hunk 1]
```diff
@@ -489,6 +490,135 @@ public function destroy($id) {
...
+					'data' => array(
+						'username' => $id,
+						'enabled' => 1
+					)
+				)
+			);
+		} else {
+			return new DataResponse(
+				array(
+					'status' => 'error',
```
Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: Please use HTTP status codes for the response instead of a status flag.

Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: Ah. You just used the same style as in the code above. Okay. Then we could keep this, but should clean this up once we touch all the code here again.

**Comment by MorrisJobke** [comment]
We already have the menu CSS in place. This is for now only implemented in the sidebar, but should be reusable: https://docs.nextcloud.org/server/9/developer_manual/app/css.html#menus

bildschirmfoto 2016-07-18 um 09 31 21


Could you check this out? And also we usually use px instead of em. Maybe this already helps to make this work.

**Comment by MorrisJobke** [comment]
But nevertheless: Thanks for your contribution :)

**Comment by Faldon** [comment]
@MorrisJobke I changed my additions to the css from em to px. There are already some declarations in em, but I would rather look into it as separate task (redesigning settings maybe, than we also can change the code for the http status codes as you mentioned above.)

The last mentioned problem (not displaying user) was a bug that seems to be present already. In the template resides a single tr for the javascript, that should be kept on reloading the user list but wasn't.

File: settings/Controller/UsersController.php [hunk 2]
```diff
@@ -245,15 +240,18 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
...
-			if($gid !== '') {
+			if($gid !== '' && $gid !== 'disabledUsers') {
 				$batch = $this->getUsersForUID($this->groupManager->displayNamesInGroup($gid, $pattern, $limit, $offset));
 			} else {
 				$batch = $this->userManager->search($pattern, $limit, $offset);
 			}
 
 			foreach ($batch as $user) {
-				$users[] = $this->formatUserForIndex($user);
+				if( ($gid!=='disabledUsers' && $user->isEnabled()) ||
```
Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: Missing spaces around `!==`

File: settings/Controller/UsersController.php [hunk 3]
```diff
@@ -266,7 +264,7 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 			$subAdminOfGroups = $gids;
 
 			// Set the $gid parameter to an empty value if the subadmin has no rights to access a specific group
-			if($gid !== '' && !in_array($gid, $subAdminOfGroups)) {
+			if($gid !== '' && $gid!== 'disabledUsers' && !in_array($gid, $subAdminOfGroups)) {
```
Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: missing space before `!==`

File: settings/Controller/UsersController.php [hunk 4]
```diff
@@ -291,7 +289,11 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 					$this->groupManager->getUserGroupIds($user),
 					$subAdminOfGroups
 				));
-				$users[] = $this->formatUserForIndex($user, $userGroups);
+				if( ($gid!=='disabledUsers' && $user->isEnabled()) ||
```
Comment (MorrisJobke) [inline settings/Controller/UsersController.php]: missing space around `!==`

File: settings/js/users/groups.js [hunk 5]
```diff
@@ -216,7 +216,7 @@ GroupList = {
 	showGroup: function (gid) {
 		GroupList.activeGID = gid;
 		UserList.empty();
-		UserList.update(gid);
+		UserList.update(gid=='_everyone'?'':gid);
```
Comment (MorrisJobke) [inline settings/js/users/groups.js]: please use `===` and spaces around the operators (`===`, `?`, `:`) for better readability.

**Comment by MorrisJobke** [comment]
Beside that it already looks quite good :)

**Comment by MorrisJobke** [comment]
For better user experience we should make the loginException extend from HintException:

```php
use OC\HintException;

class LoginException extends HintException {
}
```

@Faldon Could you add this?

@LukasReschke I don't see a risk for leaking info here. Every time the login exception is thrown it also uses a translated message.

@Faldon Beside that it works really nice 👍

**Comment by Faldon** [comment]
I'm still struggling with the code style. -.- Fixes are on their way.
@MorrisJobke Currently, the LoginController seems to not throw any Exception. Or am I looking at the wrong place?

**Comment by MorrisJobke** [comment]
> @MorrisJobke Currently, the LoginController seems to not throw any Exception. Or am I looking at the wrong place?

Correct. Then all is fine. This is the old way to implement it but it is consistent in that class. So leave it as it is :)

**Comment by MorrisJobke** [comment]
@Faldon There are conflicts that doesn't allow merging. Could you try to resolve those conflicts?

**Comment by Faldon** [comment]
Hmpf...
Merging the master branch back appearantly was not a good idea... ;-( Sorry for that. But hopefully, the merge conflicts are resolved now...

**Comment by jancborchardt** [comment]
@Faldon I just noticed the added icons. Instead please use the classes `.icon-add` and `.icon-delete` with the respective text: »Add user« or »Delete user« (or »Disable«)

**Comment by Faldon** [comment]
@jancborchardt Should I extend the icons style sheet for the disabling user icon then?

**Comment by jancborchardt** [comment]
@Faldon we can use `.icon-close` for that. It’s similar to »Delete« but not a trash can. And it’s important there’s text next to it. :)

**Comment by Faldon** [comment]
@jancborchardt Ok, I did the changes for the action menu. However, I left the code style for the "more" icon to comply with the other icons in the tr that are visible on hover.

**Comment by MorrisJobke** [comment]
The conflicts doesn't seem to be trivial. I will try to sort this out. Sorry for having this laying around that long 😢

**Comment by MorrisJobke** [comment]
Rebase on master is in #1234

@Faldon Is this okay for you? (that I take over this PR)

**Comment by MorrisJobke** [comment]
> @Faldon Is this okay for you? (that I take over this PR)

I will close this here for now.

**Comment by MorrisJobke** [comment]
Superseeded by #1234

**Pull request closed without merging**